ledger/simulation/resources.go

```diff
@@ -0,0 +1,584 @@
+		// The app local of an app and its own account is always available, so don't bother recording it.
+		return true
+	}
+	// nil map lookup is ok
+	_, ok := a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}]
+	return ok
+}
+
+func (a *GroupResourceAssignment) addLocal(addr basics.Address, aid basics.AppIndex) bool {
+	if len(a.AssetHoldings)+len(a.AppLocals) >= a.MaxCrossProductReferences {
+		return false
+	}
+	if a.AppLocals == nil {
+		a.AppLocals = make(map[ledgercore.AccountApp]struct{})
+	}
+	a.AppLocals[ledgercore.AccountApp{Address: addr, App: aid}] = struct{}{}
+	return true
+}
+
+type resourcePolicy struct {
+	assignment                  GroupResourceAssignment
+	ep                          *logic.EvalParams
+	initialBoxSurplusReadBudget *uint64
+
+	txnRootIndex   int
+	programVersion uint64
+	globalSharing  bool
+}
+
+func newResourcePolicy(ep *logic.EvalParams, groupResult *TxnGroupResult) *resourcePolicy {
+	policy := resourcePolicy{
+		assignment: makeGroupResourceAssignment(ep.TxnGroup, ep.Proto),
+		ep:         ep,
+	}
+	groupResult.UnnamedResources = &policy.assignment
+	for i := range groupResult.Txns {
+		groupResult.Txns[i].UnnamedResources = &policy.assignment.localTxnResources[i]
+	}
+	return &policy
+}
+
+func (p *resourcePolicy) AvailableAccount(addr basics.Address) bool {
+	if p.assignment.hasAccount(addr, p.ep, p.programVersion, p.globalSharing, p.txnRootIndex) {
+		return true
+	}
+	return p.assignment.addAccount(addr, p.globalSharing, p.txnRootIndex)
+}
+
+func (p *resourcePolicy) AvailableAsset(aid basics.AssetIndex) bool {
+	if p.assignment.hasAsset(aid, p.globalSharing, p.txnRootIndex) {
+		return true
+	}
+	return p.assignment.addAsset(aid, p.globalSharing, p.txnRootIndex)
+}
+
+func (p *resourcePolicy) AvailableApp(aid basics.AppIndex) bool {
+	if p.assignment.hasApp(aid, p.globalSharing, p.txnRootIndex) {
+		return true
+	}
+	return p.assignment.addApp(aid, p.ep, p.programVersion, p.globalSharing, p.txnRootIndex)
+}
+
+func (p *resourcePolicy) AllowsHolding(addr basics.Address, aid basics.AssetIndex) bool {
+	// holdings are only checked if globalSharing is true
+	if p.assignment.hasHolding(addr, aid) {
+		return true
+	}
+	return p.assignment.addHolding(addr, aid)
+}
+
+func (p *resourcePolicy) AllowsLocal(addr basics.Address, aid basics.AppIndex) bool {
+	// locals are only checked if globalSharing is true
+	if p.assignment.hasLocal(addr, aid, p.ep) {
+		return true
+	}
+	return p.assignment.addLocal(addr, aid)
+}
+
+func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string, operation logic.BoxOperation, createSize uint64) bool {
+	if p.assignment.hasBox(app, name) {
+		// We actually never expect this to happen, since the EvalContext remembers each box in
+		// order to track their dirty bytes, and it won't invoke this method if it's already seen
+		// the box.
+		return true
+	}
+	box, ok, err := p.ep.Ledger.GetBox(app, name)
+	if err != nil {
+		panic(err)
```

Comment:
Why promote this to a panic, instead of returning the err? Maybe you're trying to prevent the resourcePolicy from changing behaviour, except by failing completely?

Reply:
This function has no error return value, so I chose to panic. True, we could just return false and the evaluator will halt because it encountered an unknown box, so we'd get the same result, but the reported error would be misleading.

I don't feel strongly about this, so let me know if you have a preference